plat/kvm/arm/smp.c

```diff
@@ -0,0 +1,242 @@
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <libfdt.h>
+#include <kvm/config.h>
+#include <kvm-arm/smp.h>
+#include <kvm/intctrl.h>
+#include <arm/cpu.h>
+#include <kvm-arm/smp_defs.h>
+#include <uk/plat/io.h>
+#include <uk/plat/lcpu.h>
+#include <arm/psci.h>
+#include <ofw/fdt.h>
+#include <uk/arch/spinlock.h>
+
+const size_t arm64_cpu_size = sizeof(struct arm64_cpu);
+
+struct arm64_cpu cpus[CONFIG_UKPLAT_LCPU_MAXCOUNT];
+
+static uint64_t bsp_cpu_id;
+static unsigned int smp_numcores;
+void _lcpu_start(struct arm64_cpu *cpu);
+
+int smp_init(void)
+{
+	int fdt_cpu;
+	const fdt32_t *naddr_prop, *id_reg;
+	const struct fdt_property *cpu_prop;
+	int naddr, cell;
+	unsigned int cpu_idx;
+	int subnode;
+
+	/* Init the info for the bootstraping cpu */
+	bsp_cpu_id = ukplat_lcpu_id();
+	cpus[0].id = bsp_cpu_id;
+	cpus[0].state = LCPU_STATE_BUSY;
+	ukarch_spin_init(&cpus[0].lock);
+	uk_pr_info("Bootstrapping processor has the ID %d\n", bsp_cpu_id);
+
+	/* Init the secondary cpus vector */
+	for (cpu_idx = 1; cpu_idx < CONFIG_UKPLAT_LCPU_MAXCOUNT; cpu_idx++) {
+		cpus[cpu_idx].id = (uint64_t) -1;
+		cpus[cpu_idx].state = LCPU_STATE_OFFLINE;
+		// ukarch_spin_init(&cpus[cpu_idx].lock);
+	}
+	smp_numcores = 0;
+
+	/* Search for assigned VM cpus in DTB */
+	fdt_cpu = fdt_path_offset(_libkvmplat_cfg.dtb, "/cpus");
+	if (fdt_cpu < 0) {
+		uk_pr_err("cpus node is not found in device tree\n");
+		return -1;
+	}
+
+	/* Get address cells */
+	naddr_prop = fdt_getprop(_libkvmplat_cfg.dtb, fdt_cpu, "#address-cells", &cell);
+	if (naddr_prop == NULL) {
+		naddr = 2;
+		uk_pr_warn("Using default value of 2 for cpu address-cells.\n");
+	} else {
+		naddr = fdt32_to_cpu(naddr_prop[0]);
+	}
+	if (naddr < 0 || naddr >= FDT_MAX_NCELLS) {
```

Comment:
TODO: naddr might be uninitialized.  Can we continue if prop == NULL?

According to the [device tree specification](https://github.com/devicetree-org/devicetree-specification/releases/tag/v0.3):
	"The #address-cells and #size-cells properties are not inherited from ancestors in the devicetree.
	A DTSpec-compliant boot program shall supply #address-cells and #size-cells on all nodes that have children.
	If missing, a client program should assume a default value of 2 for #address-cells, and a value of 1 for
	#size-cells."
I think the best way to go here is setting a default value for naddr, but warn the user about the fact that it is using a default value and his device tree may be broken.